src/CowWrapper.sol

```diff
@@ -0,0 +1,176 @@
+ */
+
+/// @title CoW Protocol Authentication Interface
+/// @author CoW DAO developers
+interface ICowAuthentication {
+    /// @dev determines whether the provided address is an authenticated solver.
+    /// @param prospectiveSolver the address of prospective solver.
+    /// @return true when prospectiveSolver is an authenticated solver, otherwise false.
+    function isSolver(address prospectiveSolver) external view returns (bool);
+}
+
+/// @title CoW Protocol Settlement Interface
+/// @notice Minimal interface for CoW Protocol's settlement contract
+/// @dev Used for type-safe calls to the settlement contract's settle function
+interface ICowSettlement {
+    /// @notice Trade data structure matching GPv2Settlement
+    struct Trade {
+        uint256 sellTokenIndex;
+        uint256 buyTokenIndex;
+        address receiver;
+        uint256 sellAmount;
+        uint256 buyAmount;
+        uint32 validTo;
+        bytes32 appData;
+        uint256 feeAmount;
+        uint256 flags;
+        uint256 executedAmount;
+        bytes signature;
+    }
+
+    /// @notice Interaction data structure for pre/intra/post-settlement hooks
+    struct Interaction {
+        address target;
+        uint256 value;
+        bytes callData;
+    }
+
+    /// @notice Returns the authentication contract used by the settlement contract.
+    function authenticator() external returns (ICowAuthentication);
+
+    /// @notice Settles a batch of trades atomically
+    /// @param tokens Array of token addresses involved in the settlement
+    /// @param clearingPrices Array of clearing prices for each token
+    /// @param trades Array of trades to execute
+    /// @param interactions Array of three interaction arrays (pre, intra, post-settlement)
+    function settle(
+        address[] calldata tokens,
+        uint256[] calldata clearingPrices,
+        Trade[] calldata trades,
+        Interaction[][3] calldata interactions
+    ) external;
+}
+
+/// @title CoW Protocol Wrapper Interface
+/// @notice Interface for wrapper contracts that add custom logic around CoW settlements
+/// @dev Wrappers can be chained together to compose multiple settlement operations
+interface ICowWrapper {
+    /// @notice A human readable label for this wrapper. Used for display in explorer/analysis UIs
+    function name() external view returns (string memory);
+
+    /// @notice Initiates a wrapped settlement call
+    /// @dev This is the entry point for wrapped settlements. The wrapper will execute custom logic
+    ///      before calling the next wrapper or settlement contract in the chain.
+    /// @param settleData ABI-encoded call to ICowSettlement.settle() containing trade data
+    /// @param wrapperData Encoded data for this wrapper and the chain of next wrappers/settlement.
+    ///                    Format: [2-byte len][wrapper-specific-data][next-address]([2-byte len][wrapper-specific-data][next-address]...)
```

Comment:
Here we should make it clear that _wrapper specific data_ comes directly from the user while the  _length_ comes from the solvers. In this way people who implement wrapper know at a glance how much they can trust the data.